### PR TITLE
Add scala.Singleton trait to synthesized definitions

### LINKED ISSUE #1496: scala.Singleton is missing from scala-library-synthetics
I can't find reference to `scala.Singleton` in scala-library so I suspect it needs to be synthesized like Any/AnyRef and friends.
https://github.com/scalameta/scalameta/blob/0b41ec6a0cac00eef35fc043a74e6cffda12afdc/semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala

### PR DESCRIPTION
**Pull request by gabro**
Closes #1496

### CODE REVIEW

File: tests/jvm/src/test/resources/scalalib.expect [hunk 1]
```diff
@@ -126,3 +126,19 @@ scala.Null# => abstract final class Null.{+1 decls}
...
+Uri => scala/Singleton.class
+Text => empty
+Language => Scala
+Symbols => 2 entries
+
+Symbols:
+scala.Singleton# => final trait Singleton.{+1 decls}
+  extends Any
+scala.Singleton#`<init>`(). => primary ctor <init>: (): Singleton
+  Singleton => scala.Singleton#
```
Comment (gabro) [inline tests/jvm/src/test/resources/scalalib.expect]: not sure about this, should traits have a constructor? I suspect no, right?

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala [hunk 2]
```diff
@@ -48,11 +48,34 @@ object Scalalib {
...
+  ): ToplevelInfos = {
+    builtinClassOrTrait(props, name, bases, symbols, isTrait = true)
+  }
+
+  private def builtinClassOrTrait(
+      props: List[s.SymbolInformation.Property],
+      name: String,
+      bases: List[String],
+      symbols: List[s.SymbolInformation],
+      isTrait: Boolean): ToplevelInfos = {
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: can we replace boolean with `Kind`? That should allow us to have a single `builtin` method

File: tests/jvm/src/test/resources/scalalib.expect (hunk 1 above)
Comment (olafurpg) [reply tests/jvm/src/test/resources/scalalib.expect]: The singleton trait should not have a constructor. I think traits have synthetic constructors, but you can't reference them in code so there is no need to include it.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala (hunk 2 above)
Comment (gabro) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: 👍 sounds right. This is where I really miss union types:

```scala
private def builtinClassOrTrait(
    kind: k.CLASS | k.TRAIT,
    props: List[s.SymbolInformation.Property],
    name: String,
    bases: List[String],
    symbols: List[s.SymbolInformation]): ToplevelInfos

File: tests/jvm/src/test/resources/scalalib.expect (hunk 1 above)
Comment (gabro) [reply tests/jvm/src/test/resources/scalalib.expect]: I've removed it when it's a trait. It still has a constructor from `Any` but it's not a declared member of that trait.

File: tests/jvm/src/test/resources/scalalib.expect [hunk 3]
```diff
@@ -126,3 +126,19 @@ scala.Null# => abstract final class Null.{+1 decls}
...
+Schema => SemanticDB v3
+Uri => scala/Singleton.class
+Text => empty
+Language => Scala
+Symbols => 2 entries
+
+Symbols:
+scala.Singleton# => final trait Singleton
+  extends Any
+scala.Singleton#`<init>`(). => primary ctor <init>: (): Singleton
```
Comment (olafurpg) [inline tests/jvm/src/test/resources/scalalib.expect]: Why is the constructor here? I thought the guard against `k.CLASS` prevented traits from having constructors.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala [hunk 4]
```diff
@@ -37,18 +37,23 @@ object Scalalib {
...
+    builtin(k.CLASS, List(p.ABSTRACT, p.FINAL), "Null", List("scala.AnyRef#"), Nil)
   }
 
-  private def builtinClass(
+  def singletonTrait: ToplevelInfos = {
+    builtin(k.TRAIT, List(p.FINAL), "Singleton", List("scala.Any#"), Nil)
+  }
+
+  private def builtin(
+      kind: s.SymbolInformation.Kind,
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: I agree union type would be nice but I can live with limiting it to SymbolInformation.Kind, which is a medium sized enum

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala [hunk 5]
```diff
@@ -71,13 +76,15 @@ object Scalalib {
     )
     val builtinSig = {
       val decls = symbols.filter(_.kind == k.METHOD)
-      val tpe = s.ClassInfoType(Nil, parents, ctor.symbol +: decls.map(_.symbol))
+      val declSymbols = decls.map(_.symbol)
+      val declarations = if (kind == k.CLASS) ctor.symbol +: declSymbols else declSymbols
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: ps. `kind.isClass` works, I prefer those helpers over using `==`

File: tests/jvm/src/test/resources/scalalib.expect (hunk 3 above)
Comment (gabro) [reply tests/jvm/src/test/resources/scalalib.expect]: Not sure about this either: the interesting thing is that the `Singleton` trait doesn't have own declarations anymore (`{+1 decls}` is gone). Could that constructor be inherited from the `Any` class?

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala (hunk 5 above)
Comment (gabro) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: I'll change also `== k.METHOD` above then.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala [hunk 6]
```diff
@@ -37,18 +37,23 @@ object Scalalib {
...
   }
 
   def nullClass: ToplevelInfos = {
-    builtinClass(List(p.ABSTRACT, p.FINAL), "Null", List("scala.AnyRef#"), Nil)
+    builtin(k.CLASS, List(p.ABSTRACT, p.FINAL), "Null", List("scala.AnyRef#"), Nil)
   }
 
-  private def builtinClass(
+  def singletonTrait: ToplevelInfos = {
+    builtin(k.TRAIT, List(p.FINAL), "Singleton", List("scala.Any#"), Nil)
```
Comment (xeno-by) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: The spec says that traits can only be `SEALED`. Is `FINAL` here essential?

File: tests/jvm/src/test/resources/scalalib.expect (hunk 3 above)
Comment (xeno-by) [reply tests/jvm/src/test/resources/scalalib.expect]: I suggest we remove the constructor here. The spec says "Unlike some other metaprogramming systems for Scala, we do not create synthetic constructor symbols for traits and objects".

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala (hunk 6 above)
Comment (gabro) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalalib.scala]: Right, not essential. It was in this error message so I've applied:

```scala
scala> new scala.Singleton {}
<console>:18: error: illegal inheritance from final trait Singleton
```

File: tests/jvm/src/test/resources/scalalib.expect (hunk 3 above)
Comment (gabro) [reply tests/jvm/src/test/resources/scalalib.expect]: Sure, I had tried already, but I've missed an `if` on the top-level info symbols.

**Comment by gabro** [comment]
Thanks for the review @xeno-by, I've addressed your comments.

**Pull request merged**